libs/ml/include/ml/core/graph.hpp

```diff
@@ -675,5 +689,60 @@ std::vector<typename Graph<TensorType>::TrainablePtrType> Graph<TensorType>::Get
   return ret;
 }
 
+template <class T>
+void Graph<T>::InsertSharedCopy(std::shared_ptr<Graph<TensorType>> output_ptr)
+{
+  if (output_ptr.get() == this)
+  {
+    throw std::runtime_error("This needs to be called with a separate ptr.");
+  }
+
+  std::shared_ptr<Graph<TensorType>> copyshare = output_ptr;
+
+  // copy all the nodes, sharing the weights using MakeSharedCopy
+  for (auto const &n : nodes_)
+  {
+    std::string                       node_name = n.first;
+    std::shared_ptr<Node<TensorType>> n_ptr     = n.second;
+    OpPtrType                         op_ptr    = n_ptr->GetOp();
+
+    OpPtrType op_copyshare = op_ptr->MakeSharedCopy(op_ptr);
+
+    assert(copyshare->nodes_.find(node_name) == copyshare->nodes_.end());
+
+    copyshare->nodes_[node_name] =
+        std::make_shared<Node<TensorType>>(*n_ptr, node_name, op_copyshare);
+
+    // add node_ptr to trainable lookup etc. if required
+    auto is_trainable =
```

Comment:
pedantic but lets reserve 'is_type' for the SFINAE checks. just call these trainable_ptr and graph_ptr